### ISSUE TITLE
Storage get_url() issue

### ISSUE THREAD

**Issue by ininex** [thisbejim/Pyrebase#142]
Looks like there are a lot of 403 forbidden issues relates to Firebase storage and I think I've figured out a way to solve it:
1. Make sure to store the response after using put() for uploading images or files
2. You can check what keys are inside response, please note 'bucket' and 'downloadTokens' are important keys we are going to use in the next step
3. The actual file url that get_url() should provide is https://firebasestorage.googleapis.com/v0/b/{bucket}/o/{path_to_file}?alt=media&token={downloadTokens}
Hightlight: I think current get_url() method cannot return url correctly for storage that needs auth to access. The current method return base url appended with idToken as file download url, which is wrong: the token used here should be downloadTokens instead.

**Comment by cjbara** [comment]
The problem is that it is expecting a "Token" parameter. Insert None as the parameter to get_url and you should be fine. 

In the future, this should be defaulted to None in Pyrebase

**Comment by lohriialo** [comment]
Kindly update this in Readme doc about get_url() Could save countless hours for someone else, wasted mine today :-(

**Comment by ubihacks** [comment]
@cjbara   Inserting " None " its not  getting the  rite Download URL 
**By inserting " none" I get :**
https://firebasestorage.googleapis.com/v0/b/****/o/python%2FOct-31-2017%2FImage-12-19-26.jpg?alt=media
 **URL on Firebase for Download**
https://firebasestorage.googleapis.com/v0/b/*****/o/python%2FOct-31-2017%2FImage-12-19-26.jpg?alt=media&token=62e4e729-991a-41f1-abb9-ed8e53d65873

Any solution for this issue
Thanks

**Comment by evaverne** [comment]
Same here, how to get the token?

**Comment by wstejka** [comment]
Me too, the same issue. The authentication token is not the one which should be a part of url

**Comment by evaverne** [comment]
Does anyone know how this token forms?

**Comment by wstejka** [comment]
@evaverne I think the problem here is that in building url process there is used auth_token instead of the one generated per each file which in my opinion should be get from server. 
But if you take a look at code you'll see the request is just concatenation of configuration's value passed by user. There is no request to server to obtain token.

Here you go the get_url():


    def get_url(self, token):
        path = self.path
        self.path = None
        if path.startswith('/'):
            path = path[1:]
        if token:
            return "{0}/o/{1}?alt=media&token={2}".format(self.storage_bucket, quote(path, safe=''), token)
        return "{0}/o/{1}?alt=media".format(self.storage_bucket, quote(path, safe=''))

**Comment by evaverne** [comment]
@wstejka Exactly, that's why I thought there is an option to find out what's going on the google side. Looking through js now, it seems there is a working function for this purpose.

**Comment by ubihacks** [comment]
@thisbejim  any  solution for this

**Comment by ubihacks** [comment]
@evaverne @wstejka did you guys find out the solution

**Comment by xaiyeon** [comment]
Any luck on this?

**Comment by Pod5GS** [comment]
@xaiyeon  Does the put function return the token? Could you please tell me how to get the token? I tried but the put function return None.

**Comment by xaiyeon** [comment]
@Pod5GS Hey Pod, I'm not sure if I explain it, it would make sense... After you use the Put function you can assign it to a variable and then it actually saves the entire request into a JSON. Then you can access the object for it, specifically the downloadTokens, and you can store that into another variable and use it for the get_url token parameter. You can see my code here, it has comments, but this is for a project I head and own.

https://github.com/xaiyeon/wamain-proj/blob/master/wamain.py

**Comment by ngost** [comment]
This's not perfect solution. But, if you can give up read permission, you can do this.
-----
Change the storage rule.

-basic rule-
service firebase.storage {
  match /b/{bucket}/o {
    match /{allPaths=**} {
      allow read, write: if request.auth != null;
    }
  }
}

-changed-

service firebase.storage {
  match /b/{bucket}/o {
    match /{allPaths=**} {
      allow read;
      allow write: if request.auth != null;
    }
  }
}

**Comment by asgaraliyev** [comment]
storage.child("PHOTO_PATH").get_url(None)
pass null as a token 
It worked for me

**Comment by successanil** [comment]
> storage.child("PHOTO_PATH").get_url(None)
> pass null as a token
> It worked for me

Nice work but this will give you a url path which will be denied by google. so url is unusable.

**Comment by RiviWickramarachchi** [comment]
You can avoid this issue by simply inserting 'None' as a parameter

**Comment by bugzzbunny007** [comment]
You can pass the Parameter as "None" it Will Work for You

**Comment by hereisamara** [comment]
[](https://user-images.githubusercontent.com/91376371/162609758-4bb7f686-010c-48d3-ae34-7c626945e63b.png)[](https://user-images.githubusercontent.com/91376371/162609777-c476e244-bd43-4677-8dbe-bdf8b6b30213.png)The current get_url method takes the bearer token as a token URL parameter which in fact is a different token, named access token or download token.

I have changed the method so that it works for all types of firebase storage rules by retrieving the access token first and passing it into the URL.

Before the get_url() method returns
![image](https://user-images.githubusercontent.com/91376371/162610112-5c4eb87f-8afa-438d-88d7-ca3f37b35afb.png)

After changes, it returns the URL with the correct token
![image](https://user-images.githubusercontent.com/91376371/162610110-a64ecbaa-1c91-489f-a832-92317261d263.png)

I have opened a pull request for this fix.
https://github.com/thisbejim/Pyrebase/pull/417#issue-1198939874

**Comment by MajdMsallati** [comment]
@hereisamara but how do I apply this solution?

**Comment by hereisamara** [comment]
For now, you can use it like this. First, retrieve the access token of the file in firebase storage and pass it into the get_url method.

You will need requests module. https://pypi.org/project/requests/

![image](https://user-images.githubusercontent.com/91376371/171487943-f6de176c-07cb-4f08-bc01-fb3abdbc184c.png)

**Comment by wanazmi69** [comment]
> Looks like there are a lot of 403 forbidden issues relates to Firebase storage and I think I've figured out a way to solve it:
> 
> 1. Make sure to store the response after using put() for uploading images or files
> 2. You can check what keys are inside response, please note 'bucket' and 'downloadTokens' are important keys we are going to use in the next step
> 3. The actual file url that get_url() should provide is [https://firebasestorage.googleapis.com/v0/b/{bucket}/o/{path_to_file}?alt=media&token={downloadTokens}](https://firebasestorage.googleapis.com/v0/b/%7Bbucket%7D/o/%7Bpath_to_file%7D?alt=media&token=%7BdownloadTokens%7D)
>    Hightlight: I think current get_url() method cannot return url correctly for storage that needs auth to access. The current method return base url appended with idToken as file download url, which is wrong: the token used here should be downloadTokens instead.

**Comment by wanazmi69** [comment]
add None in get_url
example:
nike = storage.child("sepatu/nike1.jpeg").get_url(None)
